src/data/program_configuration.js

```diff
@@ -52,83 +59,143 @@ function packColor(color: Color): [number, number] {
  */
 interface Binder<T> {
     statistics: { max: number };
+    paintVertexArray?: StructArray;
 
     populatePaintArray(length: number, feature: Feature): void;
     upload(Context): void;
     destroy(): void;
 
     defines(): Array<string>;
+    isDataDriven(): boolean;
 
     setUniforms(context: Context,
                 program: Program,
                 globals: GlobalProperties,
                 currentValue: PossiblyEvaluatedPropertyValue<T>): void;
+
+    setTileSpecificUniforms(context: Context,
+                program: Program,
+                globals: GlobalProperties,
+                currentValue: PossiblyEvaluatedPropertyValue<T>,
+                tile: ?Tile,
+                crossfade: ?CrossfadeParameters): void;
 }
 
 class ConstantBinder<T> implements Binder<T> {
     value: T;
-    name: string;
+    names: Array<string>;
     type: string;
     statistics: { max: number };
+    +setTileSpecificUniforms: (Context,
+                Program,
+                GlobalProperties,
+                PossiblyEvaluatedPropertyValue<T>,
+                ?Tile) => void;
 
-    constructor(value: T, name: string, type: string) {
+
+    constructor(value: T, names: Array<string>, type: string) {
         this.value = value;
-        this.name = name;
+        this.names = names;
         this.type = type;
         this.statistics = { max: -Infinity };
     }
 
     defines() {
-        return [`#define HAS_UNIFORM_u_${this.name}`];
+        return this.names.map(name => `#define HAS_UNIFORM_u_${name}`);
+    }
+
+    isDataDriven() {
+        return false;
     }
 
     populatePaintArray() {}
     upload() {}
     destroy() {}
+    setTileSpecificUniforms() {}
 
     setUniforms(context: Context,
                 program: Program,
                 globals: GlobalProperties,
                 currentValue: PossiblyEvaluatedPropertyValue<T>) {
         const value: any = currentValue.constantOr(this.value);
         const gl = context.gl;
-        if (this.type === 'color') {
-            gl.uniform4f(program.uniforms[`u_${this.name}`], value.r, value.g, value.b, value.a);
-        } else {
-            gl.uniform1f(program.uniforms[`u_${this.name}`], value);
+        for (let i = 0; i < this.names.length; i++) {
+            const name = this.names[i];
+            if (this.type === 'color') {
+                gl.uniform4f(program.uniforms[`u_${name}`], value.r, value.g, value.b, value.a);
+            } else {
+                gl.uniform1f(program.uniforms[`u_${name}`], value);
+            }
+        }
+    }
+
+}
+
+class PatternConstantBinder<T> extends ConstantBinder<T> {
```

Comment:
I'm still inclined to say `PatternConstantBinder` should be a top-level class rather than a subclass of `ConstantBinder` (and likewise for `PatternCompositeExpressionBinder`).